### PR TITLE
[MRG]Add model status when registering model

### LINKED ISSUE #212: Add model_status Field when registering model
#### Software and Hardware Versions 
ModelCI v1.x.x, CUDA Version vx.x.x, GPU device used...

#### Problem description

Add Model status field to indicate different phases of models

    'Published' for the modified model structure
    'Training' for models under the finetune process
    'Ready' for those who finished finetuning
    'In Service' for those who are already deployed

#### Other Information 
Things you tried, stack traces, related issues, suggestions on how to fix it...

### PR DESCRIPTION
**Pull request by univerone**
Closes #212 

#### What has been done to verify that this works as intended?

Preview, after model editing
Screenshot_2021-02-06 ML ModelCI(2)

after training started
Screenshot_2021-02-06 ML ModelCI(3)

#### Before submitting this PR, please make sure you have:

- [x] run `python -m pytest tests/` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited.

### CODE REVIEW

File: modelci/experimental/finetuner/trainer.py [hunk 1]
```diff
@@ -93,12 +93,12 @@ def __init__(
     def from_training_job(cls, training_job: TrainingJob) -> 'PyTorchTrainer':
         # TODO: only support fine-tune
 
-        model_bo = ModelService.get_model_by_id(training_job.model)
-        if model_bo.engine != Engine.PYTORCH:
-            raise ValueError(f'Model engine expected `{Engine.PYTORCH}`, but got {model_bo.engine}.')
+        cls.model_bo = ModelService.get_model_by_id(training_job.model)
```
Comment (YuanmingLeee) [inline modelci/experimental/finetuner/trainer.py:96]: you shouldn't do something like this. As we have stored the model ID, you should use this model ID (by calling `self.model`) to retrieve the model object.

File: modelci/experimental/finetuner/trainer.py [hunk 2]
```diff
@@ -137,12 +137,14 @@ def from_training_job(cls, training_job: TrainingJob) -> 'PyTorchTrainer':
...
             model_train_curd.update(TrainingJobUpdate(_id=self._id, status=Status.PASS))
-            # TODO: save to database
+            # TODO: save to database and update model_status
             print(self.export_model())
 
         self._task = self._executor.submit(self.trainer_engine.fit, self.model, **self._data_loader_kwargs)
         self._task.add_done_callback(training_done_callback)
         model_train_curd.update(TrainingJobUpdate(_id=self._id, status=Status.RUNNING))
+        self.model_bo.model_status = ModelStatus.TRAINING
+        ModelService.update_model(self.model_bo)
```
Comment (YuanmingLeee) [inline modelci/experimental/finetuner/trainer.py:147]: here you can create an empty model object, and add ID and model status to it. Then, you can the the `update_model` API.

File: modelci/experimental/finetuner/trainer.py [hunk 3]
```diff
@@ -153,6 +155,8 @@ def terminate(self):
             # trigger pytorch lighting training graceful shutdown via a ^C
             self._task.set_exception(KeyboardInterrupt())
             model_train_curd.update(TrainingJobUpdate(_id=self._id, status=Status.FAIL))
+            self.model_bo.model_status = ModelStatus.PUBLISHED
+            ModelService.update_model(self.model_bo)
```
Comment (YuanmingLeee) [inline modelci/experimental/finetuner/trainer.py:159]: Here also, check the `update_model` documentation to see the usage.

File: modelci/hub/deployer/dispatcher.py [hunk 4]
```diff
@@ -98,10 +99,15 @@ def serve_by_name(args):
...
 
+    model_bo[0].model_status = ModelStatus.IN_SERVICE
+    ModelService.update_model(model_bo[0])
+
 
 def serve_by_task(args):
     model_bo = retrieve_model_by_task(task=args.task)
     serve(model_bo[0].saved_path, device=args.device, name=args.name, batch_size=args.bs)
+    model_bo[0].model_status = ModelStatus.IN_SERVICE
+    ModelService.update_model(model_bo[0])
```
Comment (YuanmingLeee) [inline modelci/hub/deployer/dispatcher.py:110]: you shall add a todo: check if the service is dispatched sucessfully

File: modelci/hub/deployer/dispatcher.py [hunk 5]
```diff
@@ -98,10 +99,15 @@ def serve_by_name(args):
     model_bo = retrieve_model(architecture_name=model, framework=framework, engine=engine)
     serve(model_bo[0].saved_path, device=args.device, name=args.name, batch_size=args.bs)
 
+    model_bo[0].model_status = ModelStatus.IN_SERVICE
+    ModelService.update_model(model_bo[0])
```
Comment (YuanmingLeee) [inline modelci/hub/deployer/dispatcher.py:103]: add a todo: check if the service is dispatched sucessfully

**Review by YuanmingLeee — COMMENTED** [review]
lgtm

**Comment by YuanmingLeee** [comment]
Please provide a screenshot to show it works

File: modelci/experimental/finetuner/trainer.py (hunk 1 above)
Comment (univerone) [reply modelci/experimental/finetuner/trainer.py:96]: The model ID hasn't been stored,  as `self.model` is `pl.LightningModule` object which stores the net instead of the model id

File: modelci/experimental/finetuner/trainer.py (hunk 2 above)
Comment (univerone) [reply modelci/experimental/finetuner/trainer.py:147]: The original model is loaded, modified, and finally saved into the database with empty `dataset` and `metric` field as a new model, then finetuning process is then operated on this new model, I don't think an extra model object should be created here.

File: modelci/experimental/finetuner/trainer.py (hunk 1 above)
Comment (univerone) [reply modelci/experimental/finetuner/trainer.py:96]: > The model ID hasn't been stored, as `self.model` is `pl.LightningModule` object which stores the net instead of the model ID

Sorry, I just find out that the model ID already is saved in `self.trainer_kwargs`

File: modelci/experimental/finetuner/trainer.py (hunk 2 above)
Comment (univerone) [reply modelci/experimental/finetuner/trainer.py:147]: As `ModelBO` object needs a few required parameters to create, maybe get the model object by id is more efficient?

File: modelci/types/vo/model_vo.py [hunk 6]
```diff
@@ -69,6 +72,13 @@ class Status(CaseInsensitiveEnum):
     FAIL = 'Fail'
 
 
+class ModelStatus(CaseInsensitiveEnum):
+    PUBLISHED = 'Published'
```
Comment (HuaizhengZhang) [inline modelci/types/vo/model_vo.py:76]: **1. I dont like the design. It is very confusing**

published means what? all models in our system can be regarded as the published

`training` means in training or `under revision`

`ready` for what?

`in service` means the service is running or just a model has been converted and bound to a serving system

**2. please add some comments**

Comment (univerone) [reply modelci/types/vo/model_vo.py:76]: PUBLISHED means the modified model is registered and wait to be trained
TRAINING means in training 
READY status means models are ready to be deployed,  newly registered pre-trained models and those finished training are marked as READY(including converted models)
IN SERVICE means models are deployed

Comment (HuaizhengZhang) [reply modelci/types/vo/model_vo.py:76]: not clear.

think of the whole workflow of modelci 

publish - convert - profiling ... edit ...

each status should be matched to one or multiple stages in chronologically order.

File: modelci/experimental/finetuner/trainer.py (hunk 1 above)
Comment (YuanmingLeee) [reply modelci/experimental/finetuner/trainer.py]: you can add an attribute `model_id` to Trainer

File: modelci/experimental/finetuner/trainer.py (hunk 2 above)
Comment (YuanmingLeee) [reply modelci/experimental/finetuner/trainer.py]: The speed is different, fetching an object from the database is considered a very slow operation compared to a memory object creation.
You can just create a model data object by setting the field to `None`, and set the id manually.
```
model_obj = ModelBO(name=None, framework=None, engine=None, version=None, dataset=None, metric=None, task=None, inputs=None, outputs=None)
model_obj._id = ...
```

File: modelci/types/vo/model_vo.py (hunk 6 above)
Comment (YuanmingLeee) [reply modelci/types/vo/model_vo.py:76]: How about using:
`draft`: edit, not trained yet
`training`: ML training
`published`: training finished, the status includes all publish - convert - profiling - deploy stages. Use the single status is because when one copy of the published model is converting, its another copy can under profiling at the same time, and another copy can be deployed.

Comment (YuanmingLeee) [reply modelci/types/vo/model_vo.py:76]: for tracking workflow stage in `publish - convert - profiling - deploy`, it should use workflow manager page

Comment (HuaizhengZhang) [reply modelci/types/vo/model_vo.py:76]: @YuanmingLeee please think of this deeply and organize a better definition like this 

image

Comment (YuanmingLeee) [reply modelci/types/vo/model_vo.py:76]: ### Workflow Status

The ModelCI workflow status values are described below:

`published`: model published to hub and ready for conversion, profiling, and deployment

`converting`: model is under conversion

`profiling`: model is under profiling

`deployed`: model is deployed as a service

`draft`: model is under edit

`validating`: model is under quick accuracy validation

`training`: model is under training

The training phase status contains: `draft` - `validating` - `training` - `published`

The serving phase status contains: `published` - `converting` - `profiling` - `deployed`

File: frontend/src/pages/Dashboard/index.tsx [hunk 7]
```diff
@@ -6,6 +6,18 @@ import reqwest from 'reqwest';
 import './index.css';
 
 const { Search } = Input;
+
+const tagColor = {
+  'Published': 'geekblue',
+  'Converted' : 'cyan',
+  'Profiling' : 'purple',
+  'In Service' : 'gold',
+  'Draft': 'lime',
```
Comment (YuanmingLeee) [inline frontend/src/pages/Dashboard/index.tsx:15]: better use red for `draft` and lime for `in service`

File: modelci/experimental/finetuner/trainer.py [hunk 8]
```diff
@@ -130,20 +132,25 @@ def from_training_job(cls, training_job: TrainingJob) -> 'PyTorchTrainer':
...
-            # TODO: save to database
+            # TODO: save to database and update model_status
             print(self.export_model())
 
         self._task = self._executor.submit(self.trainer_engine.fit, self.model, **self._data_loader_kwargs)
         self._task.add_done_callback(training_done_callback)
         model_train_curd.update(TrainingJobUpdate(_id=self._id, status=Status.RUNNING))
 
+        model_bo = ModelService.get_model_by_id(self.model_id)
+        model_bo.model_status.append(ModelStatus.TRAINING)
```
Comment (YuanmingLeee) [inline modelci/experimental/finetuner/trainer.py:151]: when the model is training, you should remove the `draft` tag, and add a `training` tag

File: modelci/experimental/finetuner/trainer.py [hunk 9]
```diff
@@ -153,6 +160,9 @@ def terminate(self):
             # trigger pytorch lighting training graceful shutdown via a ^C
             self._task.set_exception(KeyboardInterrupt())
             model_train_curd.update(TrainingJobUpdate(_id=self._id, status=Status.FAIL))
+            model_bo = ModelService.get_model_by_id(self.model_id)
+            model_bo.model_status.remove(ModelStatus.TRAINING)
```
Comment (YuanmingLeee) [inline modelci/experimental/finetuner/trainer.py:164]: remove the training tag and add a draft tag

File: modelci/hub/deployer/dispatcher.py [hunk 10]
```diff
@@ -98,10 +99,17 @@ def serve_by_name(args):
     model_bo = retrieve_model(architecture_name=model, framework=framework, engine=engine)
     serve(model_bo[0].saved_path, device=args.device, name=args.name, batch_size=args.bs)
 
+    # TODO: check if the service is dispatched sucessfully
+    model_bo[0].model_status.append(ModelStatus.IN_SERVICE)
```
Comment (YuanmingLeee) [inline modelci/hub/deployer/dispatcher.py:103]: remove published, converted tags, add in service tag

File: modelci/hub/deployer/dispatcher.py [hunk 11]
```diff
@@ -98,10 +99,17 @@ def serve_by_name(args):
...
+    ModelService.update_model(model_bo[0])
+
 
 def serve_by_task(args):
     model_bo = retrieve_model_by_task(task=args.task)
     serve(model_bo[0].saved_path, device=args.device, name=args.name, batch_size=args.bs)
+    # TODO: check if the service is dispatched sucessfully
+    model_bo[0].model_status.append(ModelStatus.IN_SERVICE)
```
Comment (YuanmingLeee) [inline modelci/hub/deployer/dispatcher.py:111]: remove published, converted tags, add in service tag

File: modelci/hub/manager.py [hunk 12]
```diff
@@ -111,7 +113,7 @@ def register_model(
 
     if convert:
         # TODO: generate from path name
-
+        # TODO: mark model status as converting
```
Comment (YuanmingLeee) [inline modelci/hub/manager.py:116]: no need to add a converting tag

File: modelci/hub/manager.py [hunk 13]
```diff
@@ -171,6 +181,10 @@ def register_model(
                 'asynchronous': False,
                 'model_info': model,
             }
+
+            model.model_status = ModelStatus.PROFILING
```
Comment (YuanmingLeee) [inline modelci/hub/manager.py:185]: remove published, converted, add profiling

**Comment by YuanmingLeee** [comment]
For the draft and training status model, it shouldn't have an engine. The engine is used for serving, but such a model is not yet ready to be served. @univerone

**Review by YuanmingLeee — APPROVED** [review]

**Review by YuanmingLeee — APPROVED** [review]
